FileIO/MPI_IO/NodePartitionedMeshReader.cpp

```diff
@@ -0,0 +1,521 @@
+    std::vector<MeshLib::Node*> const& mesh_nodes,
+    std::vector<unsigned long> const& glb_node_ids,
+    std::vector<MeshLib::Element*> const& mesh_elems) const
+{
+    return new MeshLib::NodePartitionedMesh(
+        mesh_name + std::to_string(_mpi_comm_size),
+        mesh_nodes, glb_node_ids, mesh_elems,
+        _mesh_info.regular_elements,
+        _mesh_info.global_base_nodes,
+        _mesh_info.global_nodes,
+        _mesh_info.base_nodes,
+        _mesh_info.active_base_nodes,
+        _mesh_info.active_nodes);
+}
+
+void NodePartitionedMeshReader::readElementASCII(std::ifstream &ins,
+    std::vector<unsigned long>& elem_data, const bool ghost) const
+{
+    // Set number of elements.
+    unsigned long const ne =
+        ghost ? _mesh_info.ghost_elements : _mesh_info.regular_elements;
+    unsigned long counter = ne;
+    for(unsigned long j = 0; j < ne; j++)
+    {
+        elem_data[j] = counter;
+        ins >> elem_data[counter++];  //mat. idx
+        ins >> elem_data[counter++];  //type
+        ins >> elem_data[counter];  //nnodes
+        unsigned long const nn_e =  elem_data[counter++];
+        for(unsigned long k = 0; k < nn_e; k++)
+            ins >> elem_data[counter++];
+    }
+}
+
+void NodePartitionedMeshReader::setNodes(const std::vector<NodeData> &node_data,
+    std::vector<MeshLib::Node*> &mesh_node,
+    std::vector<unsigned long> &glb_node_ids) const
+{
+    mesh_node.resize(_mesh_info.nodes);
+    glb_node_ids.resize(_mesh_info.nodes);
+
+    for(std::size_t i = 0; i < mesh_node.size(); i++)
+    {
+        NodeData const& nd = node_data[i];
+        glb_node_ids[i] = nd.index;
+        mesh_node[i] = new MeshLib::Node(nd.x, nd.y, nd.z, i);
+    }
+}
+
+void NodePartitionedMeshReader::setElements(
+    const std::vector<MeshLib::Node*> &mesh_nodes,
+    const std::vector<unsigned long> &elem_data,
+    std::vector<MeshLib::Element*> &mesh_elems, const bool ghost) const
+{
+    // Number of elements, ether ghost or regular
+    unsigned long const ne =
+        ghost ? _mesh_info.ghost_elements : _mesh_info.regular_elements;
+    unsigned long const id_offset =
+        ghost ? _mesh_info.regular_elements : 0;
+
+    for(unsigned long i = 0; i < ne; i++)
+    {
+        unsigned long counter = elem_data[i];
+
+        const unsigned mat_idx = static_cast<unsigned>( elem_data[counter++] );
+        const unsigned long e_type = elem_data[counter++];
+        unsigned long const nnodes = elem_data[counter++];
+
+        MeshLib::Node** elem_nodes = new MeshLib::Node*[nnodes];
+        for(unsigned long k = 0; k < nnodes; k++)
+            elem_nodes[k] = mesh_nodes[ elem_data[counter++] ];
+
+        MeshLib::Element* elem = nullptr;
+
+        // The element types below are defined by the mesh_partition tool
+        // available at https://github.com/ufz/mesh_partition .
+        switch(e_type)
+        {
+        case 1:
+            elem = new MeshLib::Line(elem_nodes, mat_idx);
```

Comment:
Why not ```mesh_elems[i+id_offset] = new MeshLib::Line(elem_nodes, mat_idx);```? The same for all the other elements. Then line 489 and 518 can be deleted. IMHO the readability would be the same.